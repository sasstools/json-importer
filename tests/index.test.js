const path = require('path');
const sass = require('node-sass');
const importer = require('../');

const generate = (file) => `
@import "${file}";
output { contents: inspect($${path.basename(file, '.json')}) }
`;

const compile = function(data, options = {}) {
  return new Promise((yeah, nah) => {
    return sass.render(
      Object.assign({ data, importer }, options),
      (err, results) => err ? nah(err) : yeah(results.css.toString()),
    );
  });
}

const compileSync = function(data, options = {}) {
  return new Promise((yeah, nah) => {
    try {
      const results = sass.renderSync(Object.assign({ data, importer }, options));
      yeah(results.css.toString());
    } catch (err) {
      nah(err);
    }
  });
}

describe('json-importer', () => {
  [[ 'async', compile ], [ 'sync', compileSync ]].forEach(([ label, func ]) => {
    describe(label, () => {
      it('should resolve flat json strings as Sass map', () => (
        func(generate('tests/fixtures/flat.json'))
          .then(result => expect(result).toMatchSnapshot())
      ));
      it('should resolve nested json strings as Sass map', () => (
        func(generate('tests/fixtures/nested.json'))
          .then(result => expect(result).toMatchSnapshot())
      ));
      it('should resolve json array values as Sass lists', () => (
        func(generate('tests/fixtures/list.json'))
          .then(result => expect(result).toMatchSnapshot())
      ));
      it('should not resolve Sass values to quoted strings', () => (
        func(generate('tests/fixtures/values.json'))
          .then(result => expect(result).toMatchSnapshot())
      ));
      it('should resolve json files in includePaths', () => (
        func(generate('fixtures/flat.json'), { includePaths: ['tests']})
          .then(result => expect(result).toMatchSnapshot())
      ));
    });
  });
});
