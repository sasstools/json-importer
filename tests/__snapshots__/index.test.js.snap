// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json-importer async should not resolve Sass values to quoted strings 1`] = `
"output {
  contents: (\\"short-hex-color\\": #def, \\"short-hex-color-with-alpha\\": #def0, \\"long-hex-color\\": #ddeeff, \\"long-hex-color-with-alpha\\": #ddeeff00, \\"named-color\\": \\"red\\", \\"number\\": 16, \\"number-with-percentage\\": 16%, \\"number-with-word-unit\\": 16rem, \\"decimal\\": 12.34, \\"decimal-with-percentage\\": 12.34%, \\"decimal-with-word-unit\\": 12.34rem, \\"function-call\\": black, \\"string\\": \\"foo\\", \\"string-with-spaces\\": \\"who are you\\", \\"array\\": (16, 16px, 16%, 12.34, 12.34px, 12.34%, #def, black, \\"foo\\", \\"who are you\\"), \\"map\\": (\\"short-hex-color\\": #def, \\"short-hex-color-with-alpha\\": #def0, \\"string\\": \\"foo\\", \\"string-with-spaces\\": \\"who are you\\")); }
"
`;

exports[`json-importer async should resolve flat json strings as Sass map 1`] = `
"output {
  contents: (\\"color\\": \\"red\\", \\"size\\": \\"small\\"); }
"
`;

exports[`json-importer async should resolve json array values as Sass lists 1`] = `
"output {
  contents: (\\"colors\\": \\"red green blue\\", \\"sizes\\": (\\"small\\", \\"medium\\", \\"big\\")); }
"
`;

exports[`json-importer async should resolve json files in includePaths 1`] = `
"output {
  contents: (\\"color\\": \\"red\\", \\"size\\": \\"small\\"); }
"
`;

exports[`json-importer async should resolve nested json strings as Sass map 1`] = `
"output {
  contents: (\\"colors\\": (\\"red\\": #f00, \\"blue\\": #00f), \\"sizes\\": (\\"small\\": 16px, \\"big\\": 30px)); }
"
`;

exports[`json-importer sync should not resolve Sass values to quoted strings 1`] = `
"output {
  contents: (\\"short-hex-color\\": #def, \\"short-hex-color-with-alpha\\": #def0, \\"long-hex-color\\": #ddeeff, \\"long-hex-color-with-alpha\\": #ddeeff00, \\"named-color\\": \\"red\\", \\"number\\": 16, \\"number-with-percentage\\": 16%, \\"number-with-word-unit\\": 16rem, \\"decimal\\": 12.34, \\"decimal-with-percentage\\": 12.34%, \\"decimal-with-word-unit\\": 12.34rem, \\"function-call\\": black, \\"string\\": \\"foo\\", \\"string-with-spaces\\": \\"who are you\\", \\"array\\": (16, 16px, 16%, 12.34, 12.34px, 12.34%, #def, black, \\"foo\\", \\"who are you\\"), \\"map\\": (\\"short-hex-color\\": #def, \\"short-hex-color-with-alpha\\": #def0, \\"string\\": \\"foo\\", \\"string-with-spaces\\": \\"who are you\\")); }
"
`;

exports[`json-importer sync should resolve flat json strings as Sass map 1`] = `
"output {
  contents: (\\"color\\": \\"red\\", \\"size\\": \\"small\\"); }
"
`;

exports[`json-importer sync should resolve json array values as Sass lists 1`] = `
"output {
  contents: (\\"colors\\": \\"red green blue\\", \\"sizes\\": (\\"small\\", \\"medium\\", \\"big\\")); }
"
`;

exports[`json-importer sync should resolve json files in includePaths 1`] = `
"output {
  contents: (\\"color\\": \\"red\\", \\"size\\": \\"small\\"); }
"
`;

exports[`json-importer sync should resolve nested json strings as Sass map 1`] = `
"output {
  contents: (\\"colors\\": (\\"red\\": #f00, \\"blue\\": #00f), \\"sizes\\": (\\"small\\": 16px, \\"big\\": 30px)); }
"
`;
